### PR TITLE
Get torus working with nuclear EoS

### PR DESCRIPTION
**Pull request by AstroBarker**
This PR adds necessary functionality to get the torus problem working with StellarCollapse EoS.

## PR Summary
I add the following functionalities:
- `adiabats.hpp` which includes necessary functions to compute adiabats from the table given S, Ye.
  - A unit test for this has been added.
- In the torus pgen, I pull in the various quantities needed (adiabatic S, Ye, table bounds). The calculations for `rho` and `u` are replaced by `GetStateFromEnthalpy` which returns `rho` and `u` for either the ideal gas or stellar collapse EoS.

Misc change: bug found in `root_find` where `status` was not passed through correctly and it complained when the root find failed even when status was passed in.

## TODO:
- @brryan When you have time, can you verify that this change does not break things for your applications? (Note: You may need to update `singularity-eos` to its latest and checkout `jhp/entropy`.
- I would like to unify the adiabat framework to work for the ideal gas as well, but I need to reason around this `kappa`.
- @brryan Do you know of anything immediately obvious that I would need to change for this to work with radiation? I haven't modified any of those parts of the pgen yet.
- This requires the `jhp/entropy` branch of `singularity-eos`. Once that is through I need to update the used version of singularity to latest commit. @jonahm-LANL I tried committing `singularity-eos` to latest but I think it did not stick. What is the way to do this?

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.

### CODE REVIEW

**Comment by AstroBarker** [comment]
Also the logic to make it work with IdealGas is kind of hacky.. I would like to fix this in the future by unifying the adiabat framework to work for an arbitrary EoS (without crash) but I need to better understand the `kappa` business first, and also this PR is already quite long..

**Comment by Yurlungur** [comment]
@AstroBarker looks like the singularity-eos submodule this is pinned to doesn't contain the entropy... Maybe something funny with `git submodule`s?

**Comment by Yurlungur** [comment]
> @AstroBarker looks like the singularity-eos submodule this is pinned to doesn't contain the entropy... Maybe something funny with `git submodule`s?

Oh I just saw your comment above. Not sure what's going on here. I'll chat with you on slack

**Comment by brryan** [comment]
@AstroBarker This should work fine with the existing radiative torus setup. To do that, we take in a GRMHD solution with known density/temperature everywhere, and then in each zone treat the GRMHD pressure as the total pressure and then solve the equation P_tot = P_gas(T) + P_rad(T) for T, where T will in general be less than the initial GRMHD solution's temperature. As long as the EOS used by your PR and that gas/rad partitioning step are the same (they should be of course) this should be fine. 

I will try to test for my use case soon, and at least review/approve ~tomorrow.

**Comment by Yurlungur** [comment]
@brryan we're currently facing a GPU bug with singularity-eos. Trying to track it down.

**Comment by Yurlungur** [comment]
@AstroBarker where is this as far as being able to merge it in?

**Comment by AstroBarker** [comment]
> @AstroBarker where is this as far as being able to merge it in?

I've got some local changes that I need to finish testing and push. The changes try to use the adiabat framework for the ideal gas after converting `kappa` -> `entropy`. Letting the root find go to much lower rho, T like we talked about helps but it's still a bit fickle to get `rho_rmax` the same, but is closer than before.

**Comment by Yurlungur** [comment]
> > @AstroBarker where is this as far as being able to merge it in?
> 
> I've got some local changes that I need to finish testing and push. The changes try to use the adiabat framework for the ideal gas after converting `kappa` -> `entropy`. Letting the root find go to much lower rho, T like we talked about helps but it's still a bit fickle to get `rho_rmax` the same, but is closer than before.

How far off are they from each other now?

**Comment by AstroBarker** [comment]
> > > @AstroBarker where is this as far as being able to merge it in?
> > 
> > 
> > I've got some local changes that I need to finish testing and push. The changes try to use the adiabat framework for the ideal gas after converting `kappa` -> `entropy`. Letting the root find go to much lower rho, T like we talked about helps but it's still a bit fickle to get `rho_rmax` the same, but is closer than before.
> 
> How far off are they from each other now?

Only a couple percent, so long as things are set correctly. It took some playing with e.g., how close to 0 I let the density/temp go on the adiabats and the number of samples used.

**Comment by Yurlungur** [comment]
> > > > @AstroBarker where is this as far as being able to merge it in?
> > > 
> > > 
> > > I've got some local changes that I need to finish testing and push. The changes try to use the adiabat framework for the ideal gas after converting `kappa` -> `entropy`. Letting the root find go to much lower rho, T like we talked about helps but it's still a bit fickle to get `rho_rmax` the same, but is closer than before.
> > 
> > 
> > How far off are they from each other now?
> 
> Only a couple percent, so long as things are set correctly. It took some playing with e.g., how close to 0 I let the density/temp go on the adiabats and the number of samples used.

Hmm. That's probably fine... but since the torus is a standard test problem, it's probably useful to be able to retain the old original ideal gas setup as well.

How gross would it be to include both that machinery and a runtime flag that reverts to the old analytic formula? Not set on the eos type but something like `torus/use_legacy_init` or something?

**Comment by AstroBarker** [comment]
Have elected to keep the switches on EoS in the torus init. Made a few changes to root find tolerances. I think this is good to go. There may need to be some additional work when it is time for radiation, but that can be done at that time.

**Comment by Yurlungur** [comment]
Looks like tests are failing:
https://github.com/lanl/phoebus/actions/runs/4622700920/jobs/8175695524?pr=159#step:5:344

File: inputs/torus.pin [hunk 1]
```diff
@@ -100,7 +103,7 @@ recon = weno5
 c2p_max_iter = 100
 c2p_tol = 1.e-8
 c2p_method = robust
-Ye = false
+Ye = true
```
Comment (Yurlungur) [inline inputs/torus.pin:106]: maybe still set default to false so as to not mess up the ideal gas run. I guess Ye doesn't do anything in that case.

File: src/pgen/torus.cpp [hunk 2]
```diff
@@ -183,15 +238,44 @@ void ProblemGenerator(MeshBlock *pmb, ParameterInput *pin) {
 
   RNGPool rng_pool(seed);
 
+  int nsamps = 180; // for adiabats
+  if (eos_type != "StellarCollapse") {
+    nsamps = 1;
+  }
```
Comment (Yurlungur) [inline src/pgen/torus.cpp:244]: Maybe make `nsamps` an input parameter that defaults to 180?

File: src/pgen/torus.cpp [hunk 3]
```diff
@@ -530,6 +619,39 @@ Real log_enthalpy(const Real r, const Real th, const Real a, const Real rin, con
   return lnh;
 }
 
+/**
+ * cheeky way to make this pgen work for ideal and nuclear EoS
+ * There's probably a better way to do this
+ **/
```
Comment (Yurlungur) [inline src/pgen/torus.cpp:625]: Nah this is fine.

File: src/phoebus_utils/adiabats.hpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+#include <kokkos_abstraction.hpp>
+#include <utils/error_checking.hpp>
+
+#include <spiner/databox.hpp>
+
+#include "microphysics/eos_phoebus/eos_phoebus.hpp"
+#include "phoebus_utils/root_find.hpp"
+
+// a namespace?
+using Microphysics::EOS::EOS;
```
Comment (Yurlungur) [inline src/phoebus_utils/adiabats.hpp:29]: As a general rule, it's not great to have a `using` statement outside a namespace in a header file. Maybe hide all of this behind an `Adiabats` namespace or something similar?

File: src/phoebus_utils/adiabats.hpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+template <typename EOS>
+inline void GetRhoBounds(const EOS &eos, const Real rho_min, const Real rho_max,
+                         const Real T_min, const Real T_max, const Real Ye, const Real S0,
+                         Real &lrho_min_new, Real &lrho_max_new) {
+
+  // adjust bounds slightly to avoid edges
+  const Real ADJUST = 0.01; // 0.01
```
Comment (Yurlungur) [inline src/phoebus_utils/adiabats.hpp:37]:
```suggestion
  constexpr Real ADJUST = 0.01;
```

File: src/phoebus_utils/root_find.hpp [hunk 6]
```diff
@@ -112,7 +112,7 @@ struct RootFind {
                                            RootFindStatus *status = nullptr) {
     Real ya, yb;
     refine_bracket(func, guess, a, b, ya, yb);
-    if (!check_bracket(a, b, ya, yb)) {
+    if (!check_bracket(a, b, ya, yb, status)) {
```
Comment (Yurlungur) [inline src/phoebus_utils/root_find.hpp:115]: :+1:

File: src/phoebus_utils/unit_conversions.cpp [hunk 7]
```diff
@@ -35,12 +35,17 @@ UnitConversions::UnitConversions(ParameterInput *pin) {
     int geom_mass_g_exists = pin->DoesParameterExist("units", "geom_mass_g");
     int geom_mass_msun_exists = pin->DoesParameterExist("units", "geom_mass_msun");
     int geom_length_cm_exists = pin->DoesParameterExist("units", "geom_length_cm");
+    int fluid_density_g_exists = pin->DoesParameterExist("units", "fluid_density_g");
```
Comment (Yurlungur) [inline src/phoebus_utils/unit_conversions.cpp:38]: `_g` here implies grams, rather than g/cm^3. maybe `_cgs`?

**Review by Yurlungur — APPROVED** [review]
Thanks @AstroBarker this was a fairly extensive changeset and it came out pretty clean. Nice work. Some nitpicks below, but all are non-blocking. Feel free to push back on them.

File: src/phoebus_utils/adiabats.hpp (hunk 4 above)
Comment (AstroBarker) [reply src/phoebus_utils/adiabats.hpp]: done

**Comment by AstroBarker** [comment]
Thanks for the feedback @Yurlungur -- all comments should be addressed.

File: inputs/torus.pin [hunk 8]
```diff
@@ -137,3 +147,10 @@ rin = 6.0
 rmax = 12.0
 n_inside_horizon = 5
 magnetized = true
+entropy = 4
+
+# For tabulated EoS
+<units>
+scale_free = false
+geom_mass_msun = 2.8
+fluid_density_cgs = 1.e11
```
Comment (Yurlungur) [inline inputs/torus.pin:156]: This probably will cause tests to fail since I think they're based on ideal gas torus. Maybe we need two input decks?

**Comment by Yurlungur** [comment]
Will merge as soon as tests pass. :+1:

File: inputs/torus.pin (hunk 8 above)
Comment (AstroBarker) [reply inputs/torus.pin:156]: Good catch. I meant to comment that out. I think it's probably fine to leave it as 
```
#<units>
#scale_free = false
...
```
What do you think?

Comment (Yurlungur) [reply inputs/torus.pin:156]: yeah it's fine to just comment out the nuclear EOS stuff. Better in fact, so it's easy to see how to swap between modes

Comment (AstroBarker) [reply inputs/torus.pin:156]: For that matter I meant to do the same with the `<eos>` block.

**Pull request merged**